### PR TITLE
CHN/KOR/TWN MSET Downgrade + FileCopy() -> FSFileCopy() + FBI injection code cleanup

### PR DESCRIPTION
**Pull request by DarkMatterCore**
* Removed the old FileCopy() function in favor of the new FSFileCopy() function.
* Fixed the file access modes used with f_open() in FSFileCopy().
* Simplified the FBI injection code by using goto jumps.
* Added *experimental* support for CHN/KOR/TWN MSET downgrade.
* Made the MSET downgrade code also check the TMD version.
* Updated msetdg.py Python script to add compatibility with CHN/KOR/TWN regions.

### CODE REVIEW

**Comment by DarkMatterCore** [comment]
I thoroughly tested FSFileCopy() with the file dumping/injecting feature and it works perfectly, but I first needed to fix the access modes it used. FA_READ and FA_WRITE were absent from the code, and that made rxTools freeze when opening a file.

**Comment by Syphurith** [comment]
@DarkMatterCore Well right the `FA_READ` and `FA_WRITE` could be added to FileOpen (not so.. convenient), or FSFileCopy (I forgot to add these two). Also thanks for testing it throughly.

File: rxtools/source/lib/fs.c [hunk 1]
```diff
@@ -64,9 +64,9 @@ u32 FSFileCopy(char *target, char *source) {
 	u32 blockSize = 0x4000;
 	u8 *buffer = (u8 *)0x26000200; //Temp buffer for transfer.
 	UINT byteI = 0, byteO = 0; //Bytes that read or written
-	retfatfs = f_open(&src, source, FA_OPEN_EXISTING);
+	retfatfs = f_open(&src, source, FA_READ | FA_OPEN_EXISTING);
 	if (retfatfs != FR_OK) {step = 1; goto closeExit;}
-	retfatfs = f_open(&dst, target, FA_CREATE_ALWAYS);
+	retfatfs = f_open(&dst, target, FA_WRITE | FA_CREATE_ALWAYS);
```
Comment (Syphurith) [inline rxtools/source/lib/fs.c]: Yeh thanks for these changes already.

File: release/tools/msetdg.py [hunk 2]
```diff
@@ -7,12 +7,12 @@
 
 try:
 	fileName = "msetdg.bin";
-	tid = [0x00020000, 0x00021000, 0x00022000] #JPN, USA, EUR
-	tver = [0x1A, 0x1F, 0x19]
+	tid = [0x00020000, 0x00021000, 0x00022000, 0x00026000, 0x00027000, 0x00028000] #JPN, USA, EUR, CHN, KOR, TWN
+	tver = [0x1A, 0x1F, 0x19, 0x06, 0x04, 0x05]
```
Comment (Syphurith) [inline release/tools/msetdg.py]: Thanks for your considering consoles in these regions. However it is said someone could not boot his CHN or TWN one into rxMode. Well still you can add it yup.

Comment (DarkMatterCore) [inline release/tools/msetdg.py]: Last time I checked, only CHN consoles use different ROP offsets for the MSET exploit. KOR and TWN should be fine, though, but it never hurts to check. We still need to work on the 9.x ROP chain, btw.

Comment (Syphurith) [inline release/tools/msetdg.py]: @DarkMatterCore CakesFW have get an experimental 9.x MSET to work. How about this?

Comment (DarkMatterCore) [inline release/tools/msetdg.py]: I was actually thinking about that. I'll take a look into that matter when I have more time. Right now, it's 10:30 pm and I have to get up at 4 am. haha

Comment (Syphurith) [inline release/tools/msetdg.py]: Thanks, and good night.

**Comment by kot2002** [comment]
rxTools are not working in KOR 3DS's (4.5 MSET, 4.5-9.0 SPIDER)
I had to change the region of the console to EUR to get rxTools work.

**Comment by Syphurith** [comment]
@kot2002 Thanks for information. If you have your ram dump (which could be generated with pasta or others) in your original region, you may contact someone for help. (Sorry i'm not smart enough.)

**Comment by roxas75** [comment]
Thanks fot the hard work!
@Syphurith i have to think about it though what i am doing is porting rxMode to AGB and TWL firms too, and make some content installable in the SD.
What rxTools miss is the support for MSET 4.x+.
I tested an experiment on a 6.1 system, but it actually failed.
Change the fcram address to 0x03F00000 didn't work.

**Comment by Syphurith** [comment]
@roxas75 Thanks as always. Really glad to hear your plans.

**Comment by DarkMatterCore** [comment]
I recall b1l1s saying that the encryption layer in the ROP chain had to be removed. Did you try that, @roxas75?

**Comment by roxas75** [comment]
I still have to @DarkMatterCore

**Pull request merged**